### ISSUE TITLE
Effect-polymorphic functions not CPS-converted correctly

### ISSUE THREAD

**Issue by nordlander** [actonlang/acton#181]
[This is a know flaw but we need an issue number to track it.]

The problem concerns functions that are polymorphic in their effect annotations, like
```
   my_map : X(fun: X(A)->B, arg: list[A]) -> list[B]
```
Calling this function has the same kind of side-effects as the function argument `fun` has, as witnessed by the effect variable `X` that annotates both the type of `fun` and the resulting type of `my_map`. If `my_map` is called with a pure function as a first argument, `my_map` itself will be a pure function; if the first argument is an action, `my_map` will also produce action effects, etc.

The problem occurs when the compiler CPS-converts code that includes such effect-polymorphic functions. The basic rule in the CPS pass is that functions annotated as actions are CPS-converted, while pure functions are not. Effect-polymorphic functions can play both roles, however, and should therefore also be CPS-converted in preparation for the most complex case.

That is currently not done, mainly because it also requires the CPS pass to identify and supply an identity continuation wherever an effect-polymorphic function is instantiated as pure. Our continuation passing primitives aren't sufficiently polymorphic in their return type to make this work, however, so an overhaul of these primitives is a prerequisite to fixing the CPS conversion of effect-polymorphic code.

Moreover, since calling CPS-converted functions comes with a small run-time cost, it would be beneficial if CPS conversion would not count functions with the following signature as effect-polymorphic:
```
   some_fun : X(ro) -> ty
```
where the effect variable `X` occurs neither in the argument types `ro` nor the result `ty`. Such a function could equally well be given the type
```
   some_fun : pure(ro) -> ty
```
and a translation that achieves this should preferably be inserted before the CPS pass in conjunction with fixing the above flaw.

**Comment by plajjan** [comment]
@nordlander is this related to deact or it remains?

**Comment by nordlander** [comment]
I actually disabled the ability to express effect-polymorphism altogether while working on the deactorizer. So the problem described above is indeed gone and the issue should be closed.

But I also relized quite recently that the new hierarchy of closure classes that the deactorizer work brings in also offers a perfect technique for implementing effect-polymorphic functions, should we decide we want them! But let's continue that discussion under another issue number when the need arises!
